### ISSUE TITLE
This crate is unsound

### ISSUE THREAD

**Issue by noamtashma** [y21/vec-cell#1]
Hey, this crate is unsound, because the `Clone` implementation is free to do anything. In particular, it might:
* Keep track of the original reference
* Call additional `Vec_cell` methods, even while inside another `Vec_cell` method.

The crate would be sound if you required `T: Copy` instead, since copies are guaranteed to basically "do nothing" and be equivalent to bit copies.

**Comment by y21** [comment]
Right, that's a good point, thanks for reporting it. I'll fix this soon

**Comment by y21** [comment]
Eh, this is problematic for more than just `Clone` and `s/Clone/Copy` is not enough.
There are several other methods on `VecCell` that call trait methods taking `&self`. For example, `VecCell::contains` calls `<T as PartialEq>::eq`.
This causes UB even with `T: Copy`:
```rs
#[derive(Clone, Copy, Default)]
struct Bad<'a> {
    data: i32,
    vec_cell: Option<&'a VecCell<Bad<'a>>>,
}

impl<'a> PartialEq for Bad<'a> {
    fn eq(&self, _other: &Self) -> bool {
        let inner = &self.data;

        // Force reallocation
        self.vec_cell.unwrap().resize(10000, Bad::default());

        // Use after free
        println!("{}", inner);

        false
    }
}

fn main() {
    let vec_cell = VecCell::new();
    vec_cell.push(Bad {
        data: 123,
        vec_cell: Some(&vec_cell),
    });
    vec_cell.contains(&Bad::default());
}
```

I guess one way to fix this is to either also remove these other problematic methods, *or* introduce unsafe traits for PartialEq, Ord, Clone. For that unsafe clone trait it could have a blanket impl for any `T: Copy`

**Comment by y21** [comment]
~~Nevermind, this can probably be fixed by just copying the `T` before calling `PartialEq` or `Ord`.~~
Edit: that would require either reimplementing affected methods or copying the whole slice, so this doesn't seem like a viable option... 😕
